File: lib/features/payment/domain/entities/payment.dart
/// Paiement adosse aux soldes MajiPay, execute dans MajiChrono (§11.2).
///
/// Le partage des roles est le suivant : **MajiPay tient les soldes**,
/// MajiChrono conduit la transaction. Les deux telephones s'apparient par un
/// code QR, et le mouvement est arbitre par le serveur.
///
/// Une regle domine tout le reste, et elle est portee par le code plutot que
/// par la documentation : **scanner n'autorise jamais un debit**. Le scan ne
/// fait qu'apparier deux appareils. Celui qui paie confirme toujours sur son
/// propre telephone, avec son code. Sans cette regle, quiconque scanne un code
/// affiche par un tiers pourrait se servir sur son compte.
library;

/// Sens de l'appariement.
///
/// L'argent va toujours du client au livreur ; ce qui change, c'est **qui
/// presente le code et qui le scanne**. Les deux sens existent parce que les
/// deux situations existent : un livreur qui tend son telephone au comptoir, un
/// client qui prefere garder le sien en main.
enum PaymentDirection {
  /// Le livreur presente le code, le client le scanne puis confirme.
  ///
  /// C'est une **demande d'encaissement** : elle ne debite rien tant que le
  /// client n'a pas confirme chez lui.
  collect('collect'),

  /// Le client presente le code, le livreur le scanne pour encaisser.
  ///
  /// C'est une **offre de paiement** : le client a deja confirme le montant
  /// avec son code au moment de creer l'intention. Le scan du livreur ne fait
  /// que reclamer ce qui a ete autorise.
  offer('offer');

  const PaymentDirection(this.wireName);

  final String wireName;

  static PaymentDirection fromWire(String? value) =>
      PaymentDirection.values.firstWhere(
        (d) => d.wireName == value,
        orElse: () => PaymentDirection.collect,
      );

  /// Le payeur a-t-il deja donne son accord au moment de la creation ?
  ///
  /// Vrai pour l'offre — le client a saisi son code avant d'afficher le sien.
  /// Faux pour la demande : le client n'a encore rien vu.
  bool get payerPreAuthorized => this == PaymentDirection.offer;

  /// Qui presente le code.
  bool get presentedByDriver => this == PaymentDirection.collect;
}

/// Etat d'une intention de paiement (EXI-MP05).
enum PaymentStatus {
  /// Creee, en attente d'appariement.
  pending('pending'),

  /// Appariee : le code a ete scanne, les deux appareils se connaissent.
  claimed('claimed'),

  /// Confirmee par le payeur, mouvement effectue.
  captured('captured'),

  /// Refusee, expiree, ou solde insuffisant.
  failed('failed'),

  /// Basculee en especes (EXI-MP08).
  cash('cash');

  const PaymentStatus(this.wireName);

  final String wireName;

  static PaymentStatus fromWire(String? value) =>
      PaymentStatus.values.firstWhere(
        (s) => s.wireName == value,
        orElse: () => PaymentStatus.pending,
      );

  /// Etat definitif : plus rien ne bougera, le sondage peut cesser.
  bool get isFinal =>
      this == PaymentStatus.captured ||
      this == PaymentStatus.failed ||
      this == PaymentStatus.cash;

  /// La course peut-elle etre reglee comme payee ?
  bool get settles =>
      this == PaymentStatus.captured || this == PaymentStatus.cash;
}

/// Motif d'echec, montre tel quel a l'utilisateur.
enum PaymentFailure {
  none('none'),
  insufficientFunds('insufficient_funds'),
  expired('expired'),
  declined('declined'),
  unavailable('unavailable');

  const PaymentFailure(this.wireName);

  final String wireName;

  static PaymentFailure fromWire(String? value) =>
      PaymentFailure.values.firstWhere(
        (f) => f.wireName == value,
        orElse: () => PaymentFailure.none,
      );

  /// Un repli especes est-il la suite naturelle ?
  ///
  /// Toujours, sauf quand il n'y a rien a replier. La course ne doit jamais
  /// rester bloquee sur un probleme de paiement (EXI-MP08, EXI-C43).
  bool get suggestsCash => this != PaymentFailure.none;
}

/// Solde MajiPay d'un utilisateur.
///
/// Il est **lu**, jamais calcule ni conserve : MajiPay en est la source de
/// verite. Un solde recopie dans MajiChrono divergerait au premier mouvement
/// fait ailleurs, et afficherait un montant faux au moment de payer.
class MajiPayBalance {
  const MajiPayBalance({
    required this.availableAriary,
    required this.accountRef,
    required this.fetchedAt,
  });

  final int availableAriary;

  /// Reference masquee du compte, du type `MP ** ** 4821`. Le numero complet
  /// n'a aucune raison de transiter (EXI-MP11).
  final String accountRef;

  final DateTime fetchedAt;

  bool covers(int amountAriary) => availableAriary >= amountAriary;

  static MajiPayBalance? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    return MajiPayBalance(
      availableAriary: (json['available'] as num?)?.toInt() ?? 0,
      accountRef: '${json['accountRef'] ?? ''}',
      fetchedAt:
          DateTime.tryParse('${json['fetchedAt']}')?.toLocal() ??
          DateTime.now(),
    );
  }
}

/// Intention de paiement (EXI-MP02).
///
/// Elle est **creee par le serveur**. Le mobile ne detient qu'un identifiant et
/// un jeton a usage unique : aucun secret durable ne descend sur l'appareil, et
/// un telephone vole ne donne acces a aucun solde.
class PaymentIntent {
  const PaymentIntent({
    required this.id,
    required this.deliveryId,
    required this.amountAriary,
    required this.direction,
    required this.status,
    required this.createdAt,
    required this.expiresAt,
    this.token,
    this.payerLabel,
    this.payeeLabel,
    this.failure = PaymentFailure.none,
    this.capturedAt,
    this.receiptRef,
  });

  final String id;
  final String deliveryId;
  final int amountAriary;
  final PaymentDirection direction;
  final PaymentStatus status;
  final DateTime createdAt;

  /// Peremption courte. Un code affiche sur un comptoir et oublie ne doit pas
  /// rester encaissable une heure plus tard.
  final DateTime expiresAt;

  /// Jeton a usage unique, present uniquement chez celui qui presente le code.
  /// Il n'est jamais journalise (EXI-MP11).
  final String? token;

  final String? payerLabel;
  final String? payeeLabel;
  final PaymentFailure failure;
  final DateTime? capturedAt;

  /// Reference du recu, une fois le mouvement effectue (EXI-MP10).
  final String? receiptRef;

  bool isExpiredAt(DateTime now) => !now.isBefore(expiresAt);

  /// Contenu du code QR.
  ///
  /// Il ne porte **que** l'identifiant et le jeton : ni montant, ni nom, ni
  /// numero de compte. Le scanneur interroge le serveur pour connaitre le
  /// reste, si bien qu'un code photographie a distance ne revele rien et ne
  /// vaut rien sans la confirmation du payeur.
  String get qrPayload => 'majichrono://pay?i=$id&t=${token ?? ''}';

  static PaymentIntent? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    final id = json['id'] as String?;
    if (id == null) return null;

    return PaymentIntent(
      id: id,
      deliveryId: '${json['deliveryId'] ?? ''}',
      amountAriary: (json['amount'] as num?)?.toInt() ?? 0,
      direction: PaymentDirection.fromWire(json['direction'] as String?),
      status: PaymentStatus.fromWire(json['status'] as String?),
      createdAt:
          DateTime.tryParse('${json['createdAt']}')?.toLocal() ??
          DateTime.now(),
      expiresAt:
          DateTime.tryParse('${json['expiresAt']}')?.toLocal() ??
          DateTime.now().add(PaymentQr.lifetime),
      token: json['token'] as String?,
      payerLabel: json['payerLabel'] as String?,
      payeeLabel: json['payeeLabel'] as String?,
      failure: PaymentFailure.fromWire(json['failure'] as String?),
      capturedAt: DateTime.tryParse('${json['capturedAt']}')?.toLocal(),
      receiptRef: json['receiptRef'] as String?,
    );
  }
}

/// Lecture et ecriture du contenu d'un code QR de paiement.
class PaymentQr {
  const PaymentQr._();

  /// Duree de validite d'un code. Assez pour tendre un telephone et scanner,
  /// trop courte pour qu'un code oublie serve a quoi que ce soit.
  static const Duration lifetime = Duration(minutes: 5);

  static const String scheme = 'majichrono';

  /// Analyse un contenu scanne.
  ///
  /// Retourne `null` sur tout ce qui n'est pas un code de paiement MajiChrono :
  /// une etiquette de colis, un code Wi-Fi, une publicite. Un scanneur qui
  /// accepterait n'importe quoi enverrait des identifiants inconnus au serveur.
  static ScannedPayment? parse(String? raw) {
    if (raw == null || raw.isEmpty) return null;

    final uri = Uri.tryParse(raw.trim());
    if (uri == null) return null;
    if (uri.scheme != scheme || uri.host != 'pay') return null;

    final id = uri.queryParameters['i'];
    final token = uri.queryParameters['t'];
    if (id == null || id.isEmpty) return null;
    if (token == null || token.isEmpty) return null;

    return ScannedPayment(intentId: id, token: token);
  }
}

/// Resultat de l'analyse d'un code scanne.
class ScannedPayment {
  const ScannedPayment({required this.intentId, required this.token});

  final String intentId;
  final String token;
}

/// Recu d'un retrait MajiPay (EXI-MP09).
///
/// Il porte la reference remise par le prestataire — de quoi retrouver le
/// mouvement chez MajiPay en cas de litige — et le solde **restant**, que
/// l'ecran relit sans avoir a interroger le serveur une seconde fois.
class WithdrawReceipt {
  const WithdrawReceipt({
    required this.ref,
    required this.amountAriary,
    required this.balance,
  });

  final String ref;
  final int amountAriary;
  final MajiPayBalance balance;

  static WithdrawReceipt? fromJson(Map<String, dynamic>? json) {
    if (json == null) return null;
    final balance = MajiPayBalance.fromJson(json);
    if (balance == null) return null;
    return WithdrawReceipt(
      ref: '${json['receiptRef'] ?? ''}',
      amountAriary: (json['amount'] as num?)?.toInt() ?? 0,
      balance: balance,
    );
  }
}

/// Entree du journal de paiements (§11, EXI-C39).
///
/// Un [PaymentIntent] enrichi de la position du compte courant dans la
/// transaction. La distinction porte l'affichage : un paiement **sortant**
/// (le compte est payeur) se lit en debit, un paiement **entrant** en credit.
/// Le serveur la calcule — le mobile n'a pas a comparer des identifiants.
class PaymentHistoryEntry {
  const PaymentHistoryEntry({required this.intent, required this.isOutgoing});

  final PaymentIntent intent;

  /// Vrai quand le compte courant est le payeur de la transaction.
  final bool isOutgoing;

  static PaymentHistoryEntry? fromJson(Map<String, dynamic>? json) {
    final intent = PaymentIntent.fromJson(json);
    if (intent == null) return null;
    return PaymentHistoryEntry(
      intent: intent,
      isOutgoing: json!['role'] != 'payee',
    );
  }
}

File: server/app/routers/payment.py
"""Paiement par code QR, adosse aux soldes MajiPay (§11.2).

Le mobile conduit le parcours ; **MajiPay tient l'argent**. Une regle domine
tout le reste et elle est portee par le code, pas par la documentation :
**scanner n'autorise jamais un debit**. Le scan ne fait qu'apparier deux
appareils. Celui qui paie confirme toujours sur son propre telephone. Sans cette
regle, quiconque scanne un code affiche par un tiers pourrait se servir.

Trois refus font l'essentiel de la valeur de ce routeur :

  - un jeton faux ou perime n'apparie rien ;
  - une confirmation qui ne vient pas du payeur est rejetee ;
  - une intention deja capturee n'est jamais debitee deux fois (EXI-MP06).

Payeur et beneficiaire sont **derives de la course** — expediteur et livreur
assigne — jamais d'un champ du mobile : c'est ce qui empeche un beneficiaire de
se payer lui-meme.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.deps import Idempotency, current_account, idempotency
from app.core.errors import ApiError, conflict, forbidden, not_found, unprocessable
from app.core.majipay import MajiPayError, get_gateway
from app.core.security import hash_secret, new_opaque_token, verify_secret
from app.db import get_db
from app.models import (
    Account,
    Delivery,
    PaymentDirection,
    PaymentIntent,
    PaymentStatus,
    UserRole,
)

router = APIRouter(prefix="/payments", tags=["payment"])

# Duree de validite d'un code : assez pour tendre un telephone et scanner, trop
# courte pour qu'un code oublie sur un comptoir serve une heure plus tard.
INTENT_LIFETIME = timedelta(minutes=5)


def _now() -> datetime:
    return datetime.now(timezone.utc)


class CreateIntent(BaseModel):
    deliveryId: int
    amount: int
    direction: str = PaymentDirection.collect.value
    # Le mobile transmet un role par commodite, mais le serveur ne s'y fie
    # jamais pour l'identite : elle vient du jeton et de la course.
    role: str | None = None


class ClaimBody(BaseModel):
    token: str


class ConfirmBody(BaseModel):
    role: str | None = None


class WithdrawBody(BaseModel):
    amount: int
    # Vers ou l'argent sort : un numero Mobile Money, un compte. Opaque cote
    # serveur, c'est MajiPay qui l'interprete.
    destination: str = ""


def _expire_if_needed(db: Session, intent: PaymentIntent) -> bool:
    """Perime une intention en vol dont le code a expire. Retourne vrai si oui."""
    if intent.is_final:
        return False
    if not intent.is_expired:
        return False
    intent.status = PaymentStatus.failed
    intent.failure = "expired"
    db.commit()
    return True


def _settle(db: Session, intent: PaymentIntent) -> dict:
    """Deplace l'argent chez MajiPay, une seule fois par intention.

    C'est le seul endroit ou un solde bouge. Le garde-fou est double : l'etat
    `captured` bloque un second passage cote MajiChrono, et la cle
    `settle_<id>` bloque un second mouvement cote MajiPay. L'un tient si l'autre
    est contourne.
    """
    if intent.status == PaymentStatus.captured:
        return intent.to_json()

    payer = db.get(Account, intent.payer_id)
    payee = db.get(Account, intent.payee_id)
    gateway = get_gateway(db)

    try:
        receipt = gateway.transfer(
            payer, payee, intent.amount_ariary, idem_key=f"settle_{intent.id}"
        )
    except MajiPayError as exc:
        intent.status = PaymentStatus.failed
        intent.failure = exc.failure
        db.commit()
        # 422 plutot que 500 : un solde insuffisant n'est pas une panne, c'est
        # une reponse. L'ecran propose alors le repli especes (EXI-MP08).
        raise unprocessable(exc.failure, exc.message, {"failure": exc.failure})

    intent.status = PaymentStatus.captured
    intent.captured_at = _now()
    intent.receipt_ref = receipt.ref
    db.commit()
    return intent.to_json()


@router.get("/balance")
async def balance(
    db: Session = Depends(get_db),
    account: Account = Depends(current_account),
) -> dict:
    """Solde MajiPay du compte courant.

    Il est **lu**, jamais recopie : MajiPay en est la source de verite. Un solde
    mis en cache afficherait un montant faux au moment de payer.
    """
    result = get_gateway(db).balance(account)
    db.commit()  # le bac a sable a pu provisionner la ligne du compte
    return {
        "available": result.available_ariary,
        "accountRef": result.account_ref,
        "fetchedAt": _now().isoformat(),
    }


@router.get("/history")
async def history(
    limit: int = 50,
    db: Session = Depends(get_db),
    account: Account = Depends(current_account),
) -> dict:
    """Historique des paiements du compte courant (§11, EXI-C39).

    On rend les intentions ou le compte est **payeur ou beneficiaire**, les plus
    recentes d'abord. C'est le journal que le client consulte pour retrouver un
    paiement, et le livreur pour rapprocher ses encaissements. Les intentions
    encore en vol (`pending`) en font partie : un paiement en cours a autant sa
    place dans le journal qu'un paiement solde.
    """
    capped = max(1, min(limit, 100))
    rows = (
        db.query(PaymentIntent)
        .filter(
            (PaymentIntent.payer_id == account.id)
            | (PaymentIntent.payee_id == account.id)
        )
        .order_by(PaymentIntent.created_at.desc())
        .limit(capped)
        .all()
    )
    # `role` dit au mobile de quel cote de la transaction il se trouve, sans lui
    # faire deviner en comparant des identifiants.
    items = []
    for intent in rows:
        data = intent.to_json()
        data["role"] = "payer" if intent.payer_id == account.id else "payee"
        items.append(data)
    return {"items": items}


@router.post("/intent", status_code=201)
async def create_intent(
    body: CreateIntent,
    db: Session = Depends(get_db),
    account: Account = Depends(current_account),
    idem: Idempotency = Depends(idempotency("POST /payments/intent")),
) -> dict:
    replayed = idem.replay()
    if replayed is not None:
        # Meme cle rejouee : on rend l'intention deja creee, jeton compris. Le
        # demandeur retrouve son code au lieu d'en obtenir un second, qui
        # resterait encaissable en parallele (EXI-MP06).
        return replayed[1]

    if body.amount <= 0:
        raise unprocessable("invalid_amount", "Montant invalide")

    direction = PaymentDirection(
        body.direction if body.direction in PaymentDirection._value2member_map_ else "collect"
    )

    delivery = db.get(Delivery, body.deliveryId)
    if delivery is None:
        raise not_found("Course inconnue")
    if delivery.driver_id is None:
        # Sans livreur assigne, il n'y a personne a payer. On refuse plutot que
        # de creer une intention orpheline.
        raise unprocessable("no_driver", "Aucun livreur assigne a cette course")

    payer_id, payee_id = delivery.client_id, delivery.driver_id
    if account.id not in (payer_id, payee_id):
        raise forbidden("not_a_party", "Course etrangere au compte")

    # Le sens dit qui presente le code : le livreur pour une demande
    # d'encaissement, le client pour une offre deja pre-autorisee.
    presenter = payee_id if direction == PaymentDirection.collect else payer_id
    if account.id != presenter:
        raise forbidden("wrong_presenter", "Ce n'est pas a ce compte de presenter le code")

    # Une offre est pre-autorisee par le client : on verifie le solde des la
    # creation, pour ne pas afficher un code que personne ne pourra encaisser.
    if direction == PaymentDirection.offer:
        payer = db.get(Account, payer_id)
        if get_gateway(db).balance(payer).available_ariary < body.amount:
            db.commit()
            raise unprocessable(
                "insufficient_funds",
                "Solde MajiPay insuffisant",
                {"failure": "insufficient_funds"},
            )

    token = new_opaque_token()
    intent = PaymentIntent(
        delivery_id=delivery.id,
        payer_id=payer_id,
        payee_id=payee_id,
        amount_ariary=body.amount,
        direction=direction,
        status=PaymentStatus.pending,
        token_hash=hash_secret(token),
        expires_at=_now() + INTENT_LIFETIME,
    )
    db.add(intent)
    db.commit()

    # Le jeton en clair ne sort qu'ici, et seulement a celui qui presente le
    # code. Il n'est jamais journalise (EXI-MP11).
    payload = {**intent.to_json(), "token": token}
    idem.remember(201, payload)
    return payload


@router.get("/{intent_id}")
async def read_intent(
    intent_id: str,
    db: Session = Depends(get_db),
    account: Account = Depends(current_account),
) -> dict:
    intent = db.get(PaymentIntent, intent_id)
    if intent is None:
        raise not_found("Intention inconnue")
    if account.id not in (intent.payer_id, intent.payee_id):
        raise forbidden("not_a_party", "Intention etrangere au compte")
    _expire_if_needed(db, intent)
    return intent.to_json()


@router.post("/{intent_id}/claim")
async def claim(
    intent_id: str,
    body: ClaimBody,
    db: Session = Depends(get_db),
    account: Account = Depends(current_account),
    idem: Idempotency = Depends(idempotency("POST /payments/claim")),
) -> dict:
    replayed = idem.replay()
    if replayed is not None:
        return replayed[1]

    intent = db.get(PaymentIntent, intent_id)
    if intent is None:
        raise not_found("Intention inconnue")

    # Le jeton est la seule preuve que le scanneur avait le code sous les yeux.
    # On le verifie avant tout le reste : un code faux ne doit rien apprendre.
    if not verify_secret(intent.token_hash, body.token):
        raise forbidden("bad_token", "Code invalide")

    if _expire_if_needed(db, intent):
        raise ApiError(410, "expired", "Code expire", {"failure": "expired"})

    if intent.is_final:
        # Rejouer un scan sur une intention deja reglee ne redebite rien : on
        # rend l'etat tel quel (EXI-MP06).
        return intent.to_json()

    intent.status = PaymentStatus.claimed
    db.commit()

    # Pour une offre, le payeur a deja donne son accord : la capture suit
    # immediatement. Pour une demande, on s'arrete la — le payeur devra
    # confirmer sur son propre telephone.
    if intent.direction == PaymentDirection.offer:
        body_out = _settle(db, intent)
    else:
        body_out = intent.to_json()

    idem.remember(200, body_out)
    return body_out


@router.post("/{intent_id}/confirm")
async def confirm(
    intent_id: str,
    body: ConfirmBody,
    db: Session = Depends(get_db),
    account: Account = Depends(current_account),
    idem: Idempotency = Depends(idempotency("POST /payments/confirm")),
) -> dict:
    replayed = idem.replay()
    if replayed is not None:
        return replayed[1]

    intent = db.get(PaymentIntent, intent_id)
    if intent is None:
        raise not_found("Intention inconnue")

    # Seul le payeur fait bouger l'argent. Le controle porte sur le compte en
    # base, pas sur un champ du corps : un beneficiaire ne peut pas se payer
    # lui-meme, meme si le mobile le lui demandait.
    if account.id != intent.payer_id:
        raise forbidden("not_payer", "Seul le payeur peut confirmer")

    if _expire_if_needed(db, intent):
        raise ApiError(410, "expired", "Code expire", {"failure": "expired"})

    body_out = _settle(db, intent)
    idem.remember(200, body_out)
    return body_out


@router.post("/{intent_id}/cash")
async def fallback_cash(
    intent_id: str,
    db: Session = Depends(get_db),
    account: Account = Depends(current_account),
    idem: Idempotency = Depends(idempotency("POST /payments/cash")),
) -> dict:
    replayed = idem.replay()
    if replayed is not None:
        return replayed[1]

    intent = db.get(PaymentIntent, intent_id)
    if intent is None:
        raise not_found("Intention inconnue")
    if account.id not in (intent.payer_id, intent.payee_id):
        raise forbidden("not_a_party", "Intention etrangere au compte")

    # Deja regle par MajiPay : on ne repasse pas en especes ce qui a ete debite,
    # sinon la course serait payee deux fois.
    if intent.status == PaymentStatus.captured:
        raise conflict(
            "already_captured",
            "Deja regle par MajiPay",
            {"currentState": PaymentStatus.captured.value},
        )

    intent.status = PaymentStatus.cash
    intent.captured_at = _now()
    intent.receipt_ref = f"ESP-{intent.id}"
    db.commit()

    body_out = intent.to_json()
    idem.remember(200, body_out)
    return body_out


@router.post("/withdraw")
async def withdraw(
    body: WithdrawBody,
    db: Session = Depends(get_db),
    account: Account = Depends(current_account),
    idem: Idempotency = Depends(idempotency("POST /payments/withdraw")),
) -> dict:
    """Retrait du solde MajiPay vers un moyen externe (le retrait, EXI-MP09).

    Le processus vit dans MajiChrono ; la sortie d'argent, elle, se fait chez
    MajiPay. On ne fait que la declencher et en rendre le recu. Reserve au
    livreur : c'est lui qui accumule des recettes a sortir.
    """
    replayed = idem.replay()
    if replayed is not None:
        return replayed[1]

    if account.role is not UserRole.driver:
        raise forbidden("role_forbidden", "Seul un livreur peut retirer ses gains")
    if body.amount <= 0:
        raise unprocessable("invalid_amount", "Montant de retrait invalide")

    # La cle vient de l'en-tete d'idempotence, jamais du montant : deux retraits
    # legitimes du meme montant sont deux operations distinctes, pas un doublon.
    # A defaut d'en-tete, une cle aleatoire garantit qu'on ne dedoublonne rien.
    idem_key = f"withdraw_{idem.key}" if idem.key else f"withdraw_{new_opaque_token()}"

    gateway = get_gateway(db)
    try:
        receipt = gateway.withdraw(
            account, body.amount, body.destination, idem_key=idem_key
        )
    except MajiPayError as exc:
        db.commit()
        raise unprocessable(exc.failure, exc.message, {"failure": exc.failure})

    remaining = gateway.balance(account)
    db.commit()

    body_out = {
        "receiptRef": receipt.ref,
        "amount": body.amount,
        "available": remaining.available_ariary,
        "accountRef": remaining.account_ref,
        "withdrawnAt": _now().isoformat(),
    }
    idem.remember(200, body_out)
    return body_out
